add applied fairness analysis with tune

simonpcouch opened the pull request:
An example of an applied fairness analysis with tune and yardstick based on hospital readmission data.

This is 2 of 2 applied fairness analyses I'd like to add to tidymodels.org, the first in #48.

**This is intended to be merged only _after_ the release of yardstick 1.3.0 on CRAN.** Rendered with the needed dev versions, for now.

[Here's where the new article is situated](https://deploy-preview-52--tidymodels-org.netlify.app/learn/work/fairness-readmission/) in the deploy preview.

simonpcouch committed:
add applied fairness analysis with tune

simonpcouch committed:
remove unneeded `library(bundle)`

simonpcouch committed:
remove residual plots from unnamed chunks

simonpcouch committed:
add needed pkg install

simonpcouch commented on installs.R:
```diff
@@ -47,6 +47,7 @@ packages <- c(
   "rules",
   "scales",
   "sessioninfo",
+  "simonpcouch/readmission",
```
Headed to CRAN once we feel this article is in a good spot.

simonpcouch commented on learn/work/fairness-readmission/index.md:
To be updated when dev tune + yardstick on CRAN.

simonpcouch commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,517 @@
+We can do so by first aggregating performance metrics from the `"age_bt"` result with `collect_metrics()` and then using some tidyr and dplyr for data manipulation:
+
+<!-- 
+TODO: it'd be great if `select_by_one_std_err()` could "just work" here, i.e.  
```
https://github.com/tidymodels/tune/issues/771

topepo commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,517 @@
+
+```{r}
+#| label: "load-readmission"
+#| eval: false
````
Do you mean to show the data (with the last line)? If not, I'd remove it.

juliasilge replied:
Or should this be `eval: true` to show it?

simonpcouch replied:
Ah, yes! [115a87b](https://github.com/tidymodels/tidymodels.org/pull/52/commits/115a87bc8ac9c045789fa29970a5091889296d16).

topepo commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,517 @@
+```{r}
+#| label: "load-tidymodels"
+library(tidymodels)
+library(baguette)
````
Maybe add `tidymodels_prefer()`?

simonpcouch replied:
[115a87b](https://github.com/tidymodels/tidymodels.org/pull/52/commits/115a87bc8ac9c045789fa29970a5091889296d16)!

topepo commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,517 @@
+#| label: "prop-by-race"
+#| message: false
+#| warning: false
+readmission %>%
```
I think that standard errors here are not great. They are complete driven by the magnitude of the proportions. Instead, I'd use a combination of `prop.test()` and `tidy()` to get the confidence intervals back.

simonpcouch replied:
I _think_ we're in agreement here that the demonstrated thing is a bad idea. I'm showing that, though the proportions of readmission are quite similar across `race` values, variation in the number of values being used to estimate those proportions drives greater variation of estimates across resamples. The fairness metrics are vulnerable to that variation since they measure differences in proportion in each resample. The `prop.test() %>% tidy()` approach would be one way to address that variation more sensibly, which I've written we ought to do if we decided to retain all of the factor levels, though we don't ultimately do that. 

Do let me know if I'm missing your point. :)

topepo commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,517 @@
+  readmission %>%
+  mutate(
+    race = case_when(
+      !race %in% c("Caucasian", "African American") ~ "Other",
```
omg I know it's not needed but the `!race` part is needling me.

```suggestion
      !(race %in% c("Caucasian", "African American")) ~ "Other",
```

simonpcouch replied:
Addressed in [115a87b](https://github.com/tidymodels/tidymodels.org/pull/52/commits/115a87bc8ac9c045789fa29970a5091889296d16). :)

topepo commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,517 @@
+The other preprocessor that we'll try encodes the age as a numeric variable rather than the bins as in the source data:
+
+```{r}
+#| label: "recipe-age"
````
I would add some comments here; it would be intimidating for folks that are not R experts.

topepo replied:
Also, I would add a better name for the x-axis

topepo replied:
If you are trying to show the relationship between the metrics (i.e. as ROC increses, X decreases), I would use a scatterplot matrix instead.

simonpcouch replied:
What are the comments about a plot referring to? 

[EDIT: going to run with these referring to `"plot-wflow-set-fit"`!]

topepo commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,517 @@
+-->
+
+```{r}
+#| label: "find-best-params"
````
I'm not sure that you want to introduce another concept here but this is a perfect use case for desirability functions.

simonpcouch replied:
Absolutely. f6842b0. :)

topepo reviewed (commented):
Looks great overall!

EmilHvitfeldt commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,517 @@
+      substr(2, 2) %>% 
+      as.numeric() %>%
+      `*`(10) %>% `+`(5)
+  ) %>%
```
The age thing right here isn't is a little much for a pipe in my opinion. I would pull it out as a helper function

```suggestion
# specify the age as the midpoint of the range
# rather than as an interval, like in the source data
# e.g. "[10-20]" -> 1 * 10 + 5 -> 15
get_age <- function(x) {
  res <- substr(x, 2, 2)
  res <- as.numeric(res)
  res * 10 + 5
}

recipe_age <-
  recipe(readmitted ~ ., data = readmission) %>%
  step_mutate(
    age_num = get_age(age)
  ) %>%
```

EmilHvitfeldt replied:
OR you could be overly fancy and do

```r
step_mutate(
  age_num = as.numeric(age) * 10 - 5
)
```

simonpcouch replied:
Heard! [115a87b](https://github.com/tidymodels/tidymodels.org/pull/52/commits/115a87bc8ac9c045789fa29970a5091889296d16).

juliasilge commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,517 @@
+  count(readmitted)
+```
+
+`r round(mean(readmission$readmitted == "Yes"), 3)*100`% of patients readmitted within 30 days after being discharged from the hospital. This is an example of a modeling problem with a _class imbalance_, where one value of the outcome variable is much more common than another. Now, taking a look at the counts of those in each protected class:
````
```suggestion
`r round(mean(readmission$readmitted == "Yes"), 3)*100`% of patients were readmitted within 30 days after being discharged from the hospital. This is an example of a modeling problem with a _class imbalance_, where one value of the outcome variable is much more common than another. Now, taking a look at the counts of those in each protected class:
```

simonpcouch replied:
Thank you! [115a87b](https://github.com/tidymodels/tidymodels.org/pull/52/commits/115a87bc8ac9c045789fa29970a5091889296d16). :)

juliasilge commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,517 @@
+```{r}
+#| label: "split"
+set.seed(1)
+readmission_splits <- initial_split(readmission_collapsed)
````
Want to consider using `strata` since there is some class imbalance?

simonpcouch replied:
Yes, good thought.

In [80fe7e5](https://github.com/tidymodels/tidymodels.org/pull/52/commits/80fe7e5a195413d647ea411b822a2083b39706a2) I added the argument, but it won't be reflected in the `wflow_set_fit` results until I refit the workflow set once the article is otherwise ready to go. :)

juliasilge commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,517 @@
+
+```{r}
+#| label: "resample"
+readmission_folds <- vfold_cv(readmission_train)
````
Use `strata`?

simonpcouch replied:
[80fe7e5](https://github.com/tidymodels/tidymodels.org/pull/52/commits/80fe7e5a195413d647ea411b822a2083b39706a2)!

juliasilge commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,517 @@
+
+### Metrics
+
+The metrics with which we choose to evaluate our models with are the core of our fairness analysis. In addition to the default metrics for classification in tune, `accuracy()` and `roc_auc()`, we'll compute a set of fairness metrics: `equal_opportunity()`, `equalized_odds()`, and `demographic_parity()`.
```
```suggestion
The metrics with which we choose to evaluate our models are the core of our fairness analysis. In addition to the default metrics for classification in tune, `accuracy()` and `roc_auc()`, we'll compute a set of fairness metrics: `equal_opportunity()`, `equalized_odds()`, and `demographic_parity()`.
```

simonpcouch replied:
Thank you! [115a87b](https://github.com/tidymodels/tidymodels.org/pull/52/commits/115a87bc8ac9c045789fa29970a5091889296d16).

juliasilge commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,517 @@
+
+
+::: callout-note
+The first two inputs, `accuracy()` and `roc_auc()`, are standard [yardstick](https://yardstick.tidymodels.org) metrics. The latter three are also yardstick metrics like any other, though are created using the [_metric factories_](https://yardstick.tidymodels.org/dev/reference/new_groupwise_metric.html) `equal_opportunity()`, `equalized_odds()`, and `demographic_parity()`. When passed a data-colum ...
```
This is really nice 👍

juliasilge commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,517 @@
+<!-- TODO: i'd really like to use `autoplot(wflow_set_fit, rank_metric = "roc_auc")` here, but the plot isn't usable without differentiating between the recipes. -->
+
+```{r}
+#| label: "plot-wflow-set-fit"
````
I do think this plot could be clearer, like @topepo said earlier. The code does look really overwhelming, and the x-axis is kind of awkward. I do wonder about something more like a pairs plot for all the metrics.

simonpcouch replied:
Heard. I guess the big pictures here are

1) The metrics are very correlated, and
2) Some model workflows do better on average than others.

Those could be two separate plots.🏄

simonpcouch replied:
Actually, you're spot on. Let's just plot the metrics pairwise and then print out the most performant models from `roc_auc()` in a table. :) 474ea32!

simonpcouch replied:
Unresolving, as there's an issue with the x axis labels.

simonpcouch replied:
Okay, better. :) 91ca48a.

juliasilge commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,517 @@
+
+Each dot in the plots above represents one configuration of a modeling workflow. For each modeling workflow besides logistic regressions, we've tried 10 unique combinations of hyperparameters. For the workflows using logistic regressions, we've tried only one, since we didn't mark any parameters for tuning using the `tune()` placeholder. The dots are "ranked" by `roc_auc()` in that, in each pane, ...
+
+The fairness metrics `demographic_parity()`, `equal_opportunity()`, and `equalized_odds()` all take values very close to zero for many models. Notably, the left-most models (i.e. the most performant models with respect to `roc_auc()`) happen to be among the most fair according to the metrics we've chosen.
```
Can we say more explicitly or earlier that close to zero is more fair?

simonpcouch replied:
Yes! Does a3dfdea do the trick?

juliasilge commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,517 @@
+
+```{r}
+#| label: "autoplot-age-bt"
+autoplot(wflow_set_fit, id = "age_bt")
````
Can you set fig width and height to make the metric labels more readable?

simonpcouch replied:
Yes! 5dce621. Let me know if that does the trick---the metric names are now fully visible but I'm curious how the size looks on others' setups.

simonpcouch replied:
Wrapped my head around the sizing a bit better--91ca48a.

juliasilge commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,517 @@
+From the perspective of a practitioner hoping to satisfy various stakeholders, the fact that these metrics are highly correlated makes the model selection process much easier. We can choose one fairness metric that we'd like to optimize for, and likely end up with a near-optimal configuration for the other metrics as a byproduct.
+
+:::callout-note
+In machine learning fairness, "impossibility thereoms" show that fairness definitions "are not mathematically or morally compatible in general" [@mitchell2021]. More concretely, unless we live in a world with no inequality, there is no way to satisfy many definitions of fairness at once. However, recent research emphasizes that near-fairness among more limited sets of metrics, like the three we'v ...
```
This and the previous paragraph are really nice 👍

juliasilge reviewed (commented):
This is, once again, looking fantastic! 🙌 

I left some individual comments, but my one last thought is that I think folks will come away with questions about the accuracy being basically the same as just saying "never readmit". Maybe link off to a different article on class imbalance? Or talk about adding a different metric that is more about the minority class?

simonpcouch committed:
address feedback from review, round 1

simonpcouch committed:
clarify that closer to zero -> more fair

simonpcouch committed:
set fig width + height to make metrics more readable

simonpcouch committed:
only plot metrics pairwise

simonpcouch committed:
bump figure dimensions

simonpcouch committed:
stratify on `readmitted`, link to resources

simonpcouch committed:
transition to desirability2 to select model

simonpcouch commented:
re: your general comment, @juliasilge, let me know if [80fe7e5](https://github.com/tidymodels/tidymodels.org/pull/52/commits/80fe7e5a195413d647ea411b822a2083b39706a2) does the trick. Glad to add some discussion about a metric that speaks specifically to performance in the minority class.

simonpcouch committed:
adjust `categories` to match the other article

juliasilge reviewed (approved):
This is looking great! The pairs plot especially is a big improvement.

The only last question/comment I have is whether to in fact add at least a sentence about a metric for the minority class, probably when you are discussing the results. I think it's up for debate whether that's super helpful, though; maybe instead you could link once more to the class imbalance article.

juliasilge commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,516 @@
+final_metrics <- collect_metrics(final_model)
+```
+
+The model we've selected has near-fairness with respect to the set of metrics we've chosen here. The accuracy of the model is `r round(final_metrics$.estimate[final_metrics$.metric == "accuracy"], 4)*100`%, quite similar to the accuracy that would result if the model just always predicted a patient would not readmit (`r (1 - round(mean(readmission_collapsed$readmitted == "Yes"), 4))*100`%). The ` ...
````
This is the paragraph where I think that _maybe_ there should be one more sentence either linking again to the class imbalance article or talking about a metric choice that could have given a clearer answer on minority class performance. However, use your judgment on whether it does/doesn't help that much.

simonpcouch replied:
You got it! [5293c2d](https://github.com/tidymodels/tidymodels.org/pull/52/commits/5293c2d4f3d3c1d4f486e8a8560e77a96fbb56a4).

simonpcouch committed:
note metric for minority class

EmilHvitfeldt commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,516 @@
+```
+
+::: callout-note
+We'll refer to the `race` groups in this data by their actual value (e.g. `"Caucasian"` rather than Caucasian) so as to not to take for granted the choices that the dataset authors made in choosing these categorizations. Racial categorizations are not stable across time and place---as you read on, consider how a change in the categories used in data collection might affect this analysis [@omi1994 ...
````
The series of words `so as to not to` is not my favorite

simonpcouch replied:
Ah, because there's a typo!😄 Thank you. [e21af27](https://github.com/tidymodels/tidymodels.org/pull/52/commits/e21af27668c751d5159b7f0539c321000860814f).

EmilHvitfeldt commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,516 @@
+  geom_bar()
+```
+
+While payment information on most patients is missing, most patients in this data are covered under medicare.
````
```suggestion
While payment information on most patients is missing, most patients in this data are covered under Medicare.
```

simonpcouch replied:
[e21af27](https://github.com/tidymodels/tidymodels.org/pull/52/commits/e21af27668c751d5159b7f0539c321000860814f).

EmilHvitfeldt commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,516 @@
+
+## Model Selection
+
+Now that we've evaluated a number of models with a variety of metrics, we can explore the results to determine our optimal model. Beginning by a quick exploratory plot of the distributions of our metrics:
```
```suggestion
Now that we've evaluated a number of models with a variety of metrics, we can explore the results to determine our optimal model. Beginning with a quick exploratory plot of the distributions of our metrics:
```

simonpcouch replied:
[e21af27](https://github.com/tidymodels/tidymodels.org/pull/52/commits/e21af27668c751d5159b7f0539c321000860814f).

EmilHvitfeldt commented on learn/work/fairness-readmission/index.qmd:
````diff
@@ -0,0 +1,516 @@
+autoplot(wflow_set_fit, id = "age_bt")
+```
+
+The learning rate `learn_rate` seems to have a more pronounced effect on the results metrics than the number of randomly selected predictors `mtry`. As before, we see that the most performant models with respect to `roc_auc()` tend to be the most fair with respect to our fairness metrics. Further, the values of each of the fairness metrics plotted above seem highly correlated. 
````
Minor, is it confusing that the words "with respect to" are used quite close to eachother?

simonpcouch replied:
[e21af27](https://github.com/tidymodels/tidymodels.org/pull/52/commits/e21af27668c751d5159b7f0539c321000860814f).

EmilHvitfeldt commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,516 @@
+From the perspective of a practitioner hoping to satisfy various stakeholders, the fact that these metrics are highly correlated makes the model selection process much easier. We can choose one fairness metric that we'd like to optimize for, and likely end up with a near-optimal configuration for the other metrics as a byproduct.
+
+:::callout-note
+In machine learning fairness, "impossibility thereoms" show that fairness definitions "are not mathematically or morally compatible in general" [@mitchell2021]. More concretely, unless we live in a world with no inequality, there is no way to satisfy many definitions of fairness at once. However, recent research emphasizes that near-fairness among more limited sets of metrics, like the three we'v ...
```
```suggestion
In machine learning fairness, "impossibility theorems" show that fairness definitions "are not mathematically or morally compatible in general" [@mitchell2021]. More concretely, unless we live in a world with no inequality, there is no way to satisfy many definitions of fairness at once. However, recent research emphasizes that near-fairness among more limited sets of metrics, like the three we've used here, is both possible and relevant [@bell2023].
```

simonpcouch replied:
[e21af27](https://github.com/tidymodels/tidymodels.org/pull/52/commits/e21af27668c751d5159b7f0539c321000860814f).

EmilHvitfeldt reviewed (approved):
A few notes.

It looks amazing, so excited to see this series of posts coming together!

simonpcouch committed:
address review comments

simonpcouch committed:
regenerate fits with review comments

EmilHvitfeldt commented on learn/work/fairness-readmission/index.qmd:
```diff
@@ -0,0 +1,515 @@
+#| warning: false
+readmission %>%
+  # randomly split into groups
+  mutate(., group = sample(1:10, nrow(.), replace = TRUE)) %>%
```
You should put a `set.seed()` before this code

simonpcouch replied:
Good point! [dd025fb](https://github.com/tidymodels/tidymodels.org/pull/52/commits/dd025fbef72a94c47c04159423fc97a8cb14fc54).

simonpcouch committed:
bump to cran readmission

simonpcouch committed:
set seed in code demonstrating variation in proportions

simonpcouch committed:
rm `/dev` in yardstick URL, update package versions

simonpcouch committed:
merge upstream

Merge branch 'main' into fairness-tune

# Conflicts:
#	installs.R
#	learn/index-listing.json
